libtenzir/builtins/formats/json.cpp

```diff
@@ -968,86 +975,100 @@ class default_parser final : public parser_base {
 public:
   using parser_base::parser_base;
 
-  auto parse(const chunk& json_chunk, parser_state& state)
-    -> generator<table_slice> {
+  auto parse(const chunk& json_chunk,
+             parser_state& state) -> generator<table_slice> {
     buffer_.append(
       {reinterpret_cast<const char*>(json_chunk.data()), json_chunk.size()});
     auto view = buffer_.view();
-    auto err
-      = this->parser_.iterate_many(view.data(), view.length(), max_object_size)
-          .get(stream_);
-    if (err) {
-      // For the simdjson 3.1 it seems impossible to have an error
-      // returned here so it is hard to understand if we can recover from
-      // it somehow.
-      buffer_.reset();
-      diagnostic::warning("{}", error_message(err))
-        .note("failed to parse")
-        .emit(this->ctrl_.diagnostics());
-      co_return;
-    }
-    for (auto doc_it = stream_.begin(); doc_it != stream_.end(); ++doc_it) {
-      // doc.error() will inherit all errors from *doc_it and get_value.
-      // No need to check after each operation.
-      auto doc = (*doc_it).get_value();
-      if (auto err = doc.error()) {
-        state.abort_requested = true;
-        diagnostic::error("{}", error_message(err))
-          .note("skips invalid JSON '{}'", view)
+    auto retry_capacity_failure = false;
+    do {
+      retry_capacity_failure = false;
+      auto err = this->parser_
+                   .iterate_many(view.data(), view.length(), current_batch_size)
+                   .get(stream_);
+      if (err) {
+        // For the simdjson 3.1 it seems impossible to have an error
+        // returned here so it is hard to understand if we can recover from
+        // it somehow.
+        buffer_.reset();
+        diagnostic::warning("{}", error_message(err))
+          .note("failed to parse")
           .emit(this->ctrl_.diagnostics());
         co_return;
       }
-      auto [action, slices]
-        = this->handle_selector(*doc_it, doc_it.source(), state);
-      switch (action) {
-        case parser_action::skip:
-          continue;
-        case parser_action::parse:
-          break;
-        case parser_action::yield:
-          TENZIR_ASSERT(slices);
-          for (auto& slice : *slices) {
-            co_yield std::move(slice);
+      for (auto doc_it = stream_.begin(); doc_it != stream_.end(); ++doc_it) {
+        // doc.error() will inherit all errors from *doc_it and get_value.
+        // No need to check after each operation.
+        auto doc = (*doc_it).get_value();
+        if (auto err = doc.error()) {
+          if (err == simdjson::CAPACITY) {
```

Comment:
I actually made a mistake here, because I only tried it out with the large single array file that originally caused this. In that case, it always happens on the first object (because there only is one top level object)

But in case we ever have a single >10MB object after other objects, we need to specifically address this. I've decided to just count how many objects got past the capacity issue and just skip those when retrying.

Reply:
Yes, that's what I thought. Thanks!